### PR TITLE
derp: change the protocol framing to always include a length

### PR DESCRIPTION
**Pull request by bradfitz**
Addresses one of crawshaw's TODOs.

### CODE REVIEW

File: derp/derp_client.go [hunk 1]
```diff
@@ -98,46 +105,40 @@ func (c *Client) recvServerInfo() (*serverInfo, error) {
...
+			ret = fmt.Errorf("derp.Send: %v", ret)
 		}
 	}()
 
-	if err := typeSendPacket.Write(c.bw); err != nil {
-		return err
+	// Verify len(msg) + 4 fits in a uint32 with a sanity check.
+	// This isn't a real limit, but is much higher than a IP
+	// packet size.
+	if len(msg) > 10<<20 {
```
Comment (crawshaw) [inline derp/derp_client.go]: This could be a lot smaller right? We move UDP packets, so the limit is 64kb. 1<<17?

File: derp/derp_server.go [hunk 2]
```diff
@@ -202,16 +211,13 @@ func (s *Server) verifyClient(clientKey key.Public, info *sclientInfo) error {
...
-	if err := putUint32(bw, magic); err != nil {
-		return err
-	}
-	if err := typeServerKey.Write(bw); err != nil {
-		return err
+	buf := make([]byte, 0, 40)
+	buf = append(buf, magic...)
+	buf = append(buf, s.publicKey[:]...)
+	if len(buf) != 40 {
+		panic("")
```
Comment (crawshaw) [inline derp/derp_server.go]: msg?

Comment (bradfitz) [reply derp/derp_server.go]: Just removed.

File: derp/derp_client.go (hunk 1 above)
Comment (bradfitz) [reply derp/derp_client.go]: Done. Capped and documented at 64<<10. (Let's only use shifts of 10, 20, 30 like Go)

**Pull request merged**